ocaml/gencert/selfcert.ml

```diff
@@ -0,0 +1,165 @@
+
+let defer f = Fun.protect ~finally:f
+
+(** [write_cert] writes a PEM file to [path]. It attempts to do that
+  atomically by writing to a temporary file in the same directory first and
+  renaming the file at the end *)
+let write_certs path key cert =
+  let delim = "\n" in
+  let temp_dir = Filename.dirname path in
+  let tmp = Filename.temp_file ~temp_dir "certify-" ".tmp" in
+  try
+    let fd = Unix.openfile tmp [Unix.O_WRONLY] 0o600 in
+    defer (fun () -> Unix.close fd) @@ fun () ->
+    UX.really_write fd (Cstruct.to_string key) 0 (Cstruct.len key) ;
+    UX.really_write_string fd delim ;
+    UX.really_write fd (Cstruct.to_string cert) 0 (Cstruct.len cert) ;
+    Unix.rename tmp path ;
+    R.ok ()
+  with e -> R.error_msg (Printexc.to_string e)
+
+let expire_in days =
+  let seconds = days * 24 * 60 * 60 in
+  let start = Ptime_clock.now () in
+  match Ptime.(add_span start @@ Span.of_int_s seconds) with
+  | Some expire ->
+      R.ok (start, expire)
+  | None ->
+      R.error_msgf "can't represent %d as time span" days
+
+let add_dns_names extension = function
+  | [] ->
+      extension
+  | names ->
+      X509.Extension.(
+        add Subject_alt_name
+          (false, X509.General_name.(singleton DNS names))
+          extension)
+
+let sign days key pubkey issuer req alt_names =
+  expire_in days >>= fun (valid_from, valid_until) ->
+  match (key, pubkey) with
+  | `RSA priv, `RSA pub when Rsa.pub_of_priv priv = pub ->
+      let extensions = add_dns_names X509.Extension.empty alt_names in
+      X509.Signing_request.sign ~valid_from ~valid_until ~extensions req key
+        issuer
+      |> R.reword_error (fun _ -> Printf.sprintf "signing failed" |> R.msg)
+  | _ ->
+      R.error_msgf "public/private keys don't match (%s)" __LOC__
+
+(** call openssl and return stdout, stderr as strings *)
+let call_openssl args =
+  let openssl = !Constants.openssl_path in
+  let home =
+    match Sys.getenv_opt "HOME" with
+    | None ->
+        D.warn
+          "environment variable 'HOME' is unavailable, falling back to \
+           HOME=/root" ;
+        "/root"
+    | Some path ->
+        path
+  in
+  let env =
+    [|"PATH=" ^ String.concat ":" Forkhelpers.default_path; "HOME=" ^ home|]
+  in
+  Forkhelpers.execute_command_get_output openssl ~env args
+
+(** [generate_private_key] calls openssl to generate an RSA key of
+  [length] bits. *)
+let generate_private_key length =
+  let args = ["genrsa"; string_of_int length] in
+  let stdout, _stderr = call_openssl args in
+  stdout
+
+let selfsign name alt_names length days certfile =
+  let rsa =
+    try
+      generate_private_key length
+      |> Cstruct.of_string
+      |> X509.Private_key.decode_pem
+      |> R.get_ok
+      |> function
+      | `RSA x ->
+          x
+    with e ->
+      D.error "generating RSA key for %s failed: %s" certfile
+        (Printexc.to_string e) ;
+      failwith (Printf.sprintf "generating RSA key for %s failed" certfile)
```

Comment:
Would the gencert binary log correctly the error message? This could be used instead to make sure it's shown everywhere
```suggestion
      let err_msg =
        Printf.sprintf "generating RSA key for %s failed: %s" certfile
          (Printexc.to_string e)
      in
      D.error "%s" err_msg ; failwith err_msg
```